examples/llm_plus_gnn/glem.py

```diff
@@ -0,0 +1,372 @@
+    idx_to_select = torch.randperm(test_idx.numel())[:num_pseudo_labels]
+    pseudo_labels_idx = test_idx[idx_to_select]
+    train_idx = torch.cat(
+        (gold_idx, pseudo_labels_idx))  # augmented train_indx
+
+    print(f'train_idx: {train_idx.size(0)}, '
+          f'gold_idx: {gold_idx.size(0)}, '
+          f'pseudo labels ratio: {pl_ratio}, '
+          f'{train_idx.size(0)/gold_idx.size(0) - 1.0}')
+    gold_dataset = torch.utils.data.Subset(dataset=text_dataset,
+                                           indices=gold_idx)
+    train_dataset = torch.utils.data.Subset(dataset=text_dataset,
+                                            indices=train_idx)
+    # valid_dataset = torch.utils.data.Subset(dataset=text_dataset,
+    #                                         indices=valid_idx)
+
+    # ========================== LM Data Loader ===============================
+
+    print('Building language model dataloader...', end='-->')
+    from torch_geometric.loader import DataLoader
+    text_pretrain_loader = DataLoader(gold_dataset, batch_size=lm_batch_size,
+                                      drop_last=False, pin_memory=False,
+                                      shuffle=True)
+    text_train_loader = DataLoader(train_dataset, batch_size=lm_batch_size,
+                                   drop_last=False, pin_memory=False,
+                                   shuffle=True)
+    # text_val_loader = DataLoader(valid_dataset, batch_size=lm_batch_size * 4,
+    #                              drop_last=False, pin_memory=True,
+    #                              shuffle=False)
+    text_data_loader = DataLoader(text_dataset, batch_size=lm_batch_size * 4,
+                                  drop_last=False, pin_memory=False,
+                                  shuffle=False)
+    print('done')
+
+    # =========================== GNN Data Loader =============================
+    initial_memory = torch.cuda.memory_allocated()
+    data = data.to(device)
+
+    current_memory_1 = torch.cuda.max_memory_allocated()
+    gpu_usage = float(current_memory_1 - initial_memory) / 1073741824
+    # Print the maximum memory usage after running the model
+    print(f'GPU memory usage -- data to gpu: {gpu_usage:.2f} GB')
+
+    print('build GNN dataloader(GraphSAGE NeighborLoader)', end='-->')
+    from torch_geometric.loader import NeighborLoader
+
+    # train on gold data w/o pseudo labels
+    graph_pretrain_loader = NeighborLoader(
+        data,
+        input_nodes=gold_idx,
+        num_neighbors=[15, 10, 5],
+        batch_size=gnn_batch_size,
+        shuffle=True,
+        num_workers=12,
+        persistent_workers=True,
+    )
+
+    # graph data loader w/ pseudo labels in M-step
+    graph_train_loader = NeighborLoader(
+        data,
+        input_nodes=train_idx,
+        num_neighbors=[15, 10, 5],
+        batch_size=gnn_batch_size,
+        shuffle=True,
+        num_workers=12,
+        persistent_workers=True,
```

Comment:
im curious what happens if you set persistent_workers=False, please test it and report the outcome (change in speed etc) so that we can decide the best default

Reply:
test gnn training 50 epochs w/ full inference on ogbn-arxiv, running time 32 seconds when `persistent_workers=True` --> 85 seconds when `persistent_workers=False`, memory usage not changing much

Reply:
okay good, lets keep True for now